ext/exif/exif.c

```diff
@@ -1323,7 +1323,7 @@ static const maker_note_type maker_note_array[] = {
 
 static HashTable *exif_make_tag_ht(tag_info_type *tag_table)
 {
-	HashTable *ht = malloc(sizeof(HashTable));
+	HashTable *ht = pmalloc(sizeof(HashTable));
```

Comment:
Can we take the opportunity of the churn to also avoid repeating the type unnecessarily?

```suggestion
	HashTable *ht = pmalloc(sizeof(*ht));
```

------------

see: https://github.com/haproxy/haproxy/blob/master/dev/coccinelle/xalloc_size.cocci

Reply:
We are not going to change all such usage so I would prefer to stick with full types. I find it also more readable.

Reply:
> We are not going to change all such usage

What do you mean by “we are not going to”?

Reply:
> I'm not sure if you meant doing just for pmalloc of changing every single sizeof.

In cases where spelling out the type explicitly is redundant and prone to errors due to the `sizeof()` and the actual type going out of sync during refactoring. This would be:

- All variants of `malloc()`.
- `memset()`.

> If latter, you don't get full agreement as it would be a big churn

I would make the changes in an automated fashion using Coccinelle, which makes them easy to review and easy to reproduce. I do not expect this to cause many conflicts, since it's unlikely for an in-progress patch to touch these locations.

Timewise, the time of branching 8.5 would be good for this type of bulk change, so that they happened before development of 8.6 starts. I've already planned to solve the `int/bool` conversion in bulk then (cc @nielsdos).

Reply:
> There's a non-trivial amount of time spent in PRs discussing the same nits (e.g. prefer `/* */` over `//`, though I have no idea whether these kinds of things are solvable with Coccinelle).

That example would probably rather be clang-format. I believe Coccinelle can also deal with comments, but the intended use case is making changes to the code based on type and control-flow information with a convenient syntax that resembles a patch file and thus is easy to grok even for less experienced users.

To use `malloc()` as an example:

```
@@
type T;
T *x;
@@

  x = malloc(
- sizeof(T)
+ sizeof(*x)
  )
```

This defines placeholders of an arbitrary type `T` and a variable `x` that is a pointer to `T`. It then searches for constructs `x = malloc(sizeof(T))` and replaces the `sizeof(T)` by `sizeof(*x)`.

For a C program:

```
#include <stdlib.h>

int
main(void) {
	int *foo = malloc(sizeof(int));
	int *bar = malloc(sizeof(double));
}
```

running that Coccinelle patch would make the following changes:

```
$ spatch -sp_file example.cocci test.c 
init_defs_builtins: /usr/lib/coccinelle/standard.h
HANDLING: test.c
diff = 
--- test.c
+++ /tmp/cocci-output-4310-cbc9b9-test.c
@@ -2,6 +2,6 @@
 
 int
 main(void) {
-	int *foo = malloc(sizeof(int));
+	int *foo = malloc(sizeof(*foo));
 	int *bar = malloc(sizeof(double));
 }
```

If we now wanted to find places where the types differ, we could use the following patch:

```
@@
type T;
type U;
T *x;
@@

  x = malloc(
(
  sizeof(T)
|
  sizeof(*x)
|
* sizeof(U)
)
  )
```

The `*` is an indicator to just “highlight” the match and not perform any replacement. We define an additional type `U` and then define multiple alternatives. For the correct variants `sizeof(T)` and `sizeof(*x)` nothing is reported. For `sizeof(U)` (where U != T, since T was already handled), a report is created:

```
$ spatch -sp_file example.cocci test.c 
init_defs_builtins: /usr/lib/coccinelle/standard.h
HANDLING: test.c
diff = 
--- test.c
+++ /tmp/cocci-output-4315-fa4dae-test.c
@@ -3,5 +3,4 @@
 int
 main(void) {
 	int *foo = malloc(sizeof(int));
-	int *bar = malloc(sizeof(double));
 }
```

Reply:
Coccinelle works very well and is easy to use. I've used it a few times and it's been battle tested against the Linux kernel, which isn't particularly known for it's consistent coding style and is known for it's complex macros etc.
I'm in favor to use the malloc sizeof pattern Tim proposes, and if we don't want to convert existing code (which I think we should), we should at least use that pattern for _new_ code.